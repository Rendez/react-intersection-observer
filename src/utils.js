const marginRE = /^-?\d*\.?\d+(px|%)$/;

export function parseRootMargin(rootMargin) {
    const marginString = rootMargin ? rootMargin.trim() : '0px';
    const [m0 = '0px', m1 = m0, m2 = m0, m3 = m1] = marginString.split(/\s+/).map(margin => {
        if (!marginRE.test(margin)) {
            throw new Error('rootMargin must be a string literal containing pixels and/or percent values');
        }
        return margin;
    });

    return `${m0} ${m1} ${m2} ${m3}`;
}

export function shallowCompare(next, prev) {
    if (Array.isArray(next) && Array.isArray(prev)) {
        if (next.length === prev.length) {
            return next.some((_, index) => shallowCompare(next[index], prev[index]));
        }
    }
    return next !== prev;
}
