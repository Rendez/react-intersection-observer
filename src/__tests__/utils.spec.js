/* eslint-env jest */
import { parseRootMargin, shallowCompare } from '../utils';

describe('parseRootMargin', () => {
    test('throws when using wrong units', () => {
        expect(() => parseRootMargin('10')).toThrowErrorMatchingInlineSnapshot(
            `"rootMargin must be a string literal containing pixels and/or percent values"`,
        );
        expect(() => parseRootMargin('10% 10')).toThrowErrorMatchingInlineSnapshot(
            `"rootMargin must be a string literal containing pixels and/or percent values"`,
        );
    });

    test('returns rootMargins with all four values', () => {
        expect(parseRootMargin()).toBe('0px 0px 0px 0px');
        expect(parseRootMargin(null)).toBe('0px 0px 0px 0px');
        expect(parseRootMargin('')).toBe('0px 0px 0px 0px');
        expect(parseRootMargin('10px 5px 0%')).toBe('10px 5px 0% 5px');
        expect(parseRootMargin('10px  ')).toBe('10px 10px 10px 10px');
        expect(parseRootMargin(' 10px 5px')).toBe('10px 5px 10px 5px');
        expect(parseRootMargin('10px 5px  0% 1%')).toBe('10px 5px 0% 1%');
    });
});

describe('shallowCompare', () => {
    const comparerFn = (nextProps, prevProps) =>
        ['disabled', 'root', 'rootMargin', 'threshold'].some(option =>
            shallowCompare(nextProps[option], prevProps[option]),
        );

    test('should return true if threshold array length is not the same', () => {
        const nextProps = { threshold: [0.25, 0.5] };
        const prevProps = { threshold: [0.25, 0.5, 0.75] };

        expect(comparerFn(nextProps, prevProps)).toBeTruthy();
    });

    test('should return true if threshold array length is the same but not equal', () => {
        const nextProps = { threshold: [0.25, 0.75, 0.5] };
        const prevProps = { threshold: [0.25, 0.5, 0.75] };

        expect(comparerFn(nextProps, prevProps)).toBeTruthy();
    });

    test('should return false if options are equal', () => {
        const nextProps = { disabled: true, root: 1, rootMargin: 2, threshold: [0.25, 0.75, 0.5] };
        const prevProps = { ...nextProps };

        expect(comparerFn(nextProps, prevProps)).toBeFalsy();
    });

    test('should return true if options are different', () => {
        const nextProps = { disabled: true, root: 1, rootMargin: 2, threshold: [0.25, 0.75, 0.5] };
        const prevProps = { ...nextProps, threshold: 1 };

        expect(comparerFn(nextProps, prevProps)).toBeTruthy();
    });
});
