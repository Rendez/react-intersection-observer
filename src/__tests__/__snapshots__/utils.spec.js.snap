// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRootMargin throws when using wrong type 1`] = `"rootMargin must be a String"`;

exports[`parseRootMargin throws when using wrong type 2`] = `"rootMargin must be a String"`;

exports[`parseRootMargin throws when using wrong units 1`] = `"rootMargin must be specified in pixels or percent"`;

exports[`parseRootMargin throws when using wrong units 2`] = `"rootMargin must be specified in pixels or percent"`;
